### PR TITLE
add support for array input 

### PR DESCRIPTION
**Pull request by stevenbrix**
This PR adds support for arrays as input

## Changes

1. Add `Array+ToAbi` and `Array+FromABI` methods. These use the `WinRTBridgeable`, `AbiBridge` and `AbiInterfaceBridge` to convert swift arrays to winrt and vice-versa. (FromAbi is needed for VTable implementations). These methods are what we use in Arc today when handwriting these APIs
2. make necessary code changes in code_writers.h to support array input parameters
3. added tests, moved test methods from `Class` to `ArrayMethods`
4. moved common test types to a `Types.idl`

Fixes DEVEX-32

### CODE REVIEW

File: tests/test_component/cpp/ArrayMethods.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+    private:
+
+        static hstring InEnum(winrt::test_component::Signed const& value);
+
+        template<typename T>
+        static void simulate_rpc_behavior(array_view<T> const& value)
+        {
+            // RPC requires array pointers to be non-null.
+            if (value.begin() == nullptr)
```
Comment (tristanlabelle) [inline tests/test_component/cpp/ArrayMethods.h:26]: Curious, where in WinRT is this important?

File: tests/test_component/cpp/ArrayMethods.cpp [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    {
+        switch (value)
+        {
+        case Signed::First: return L"First";
+        case Signed::Second: return L"Second";
+        case Signed::Third: return L"Third";
+        }
+
+        throw hresult_invalid_argument();
+    }
```
Comment (tristanlabelle) [inline tests/test_component/cpp/ArrayMethods.cpp:112]: This looks like it shouldn't live here?

**Review by tristanlabelle — APPROVED** [review]
I reviewed all but `code_writers.h`, which I struggle to follow

File: tests/test_component/cpp/ArrayMethods.cpp (hunk 2 above)
Comment (stevenbrix) [reply tests/test_component/cpp/ArrayMethods.cpp:112]: it was in the class which shared all of the implementations, but needed to be duplicated to remove array tests from that large file. better to keep them separate as it's for this specific test then share i think

File: swiftwinrt/code_writers/can_write.h [hunk 3]
```diff
@@ -31,8 +31,8 @@ namespace swiftwinrt
             auto param_name = get_swift_name(param);
 
             // TODO: WIN-32 swiftwinrt: support array types
```
[the remark this answers is not in the document]
Comment (stevenbrix) [reply swiftwinrt/code_writers/can_write.h:33]: not fully, a follow up pr will add support for output params

File: swiftwinrt/Resources/Support/Array+FromAbi.swift [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+@_spi(WinRTInternal)
+extension Array where Element: Numeric {
+    public static func from(abi: WinRTArrayAbi<Element>) -> [Element] {
+        Array(UnsafeBufferPointer(start: abi.start, count: Int(abi.count)))
+    }
+}
+
+@_spi(WinRTInternal)
+extension Array where Element: RawRepresentable, Element.RawValue: Numeric {
```
[the remark this answers is not in the document]
Comment (stevenbrix) [reply swiftwinrt/Resources/Support/Array+FromAbi.swift:22]: it covers enums as they are currently written

File: tests/test_component/cpp/ArrayMethods.h (hunk 1 above)
Comment (stevenbrix) [reply tests/test_component/cpp/ArrayMethods.h:26]: this was an artifact of using the same test component that C++/WinRT used. i imagine from the comment that it's when arrays are marshaled cross-proc via RPC

File: swiftwinrt/code_writers/can_write.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (tristanlabelle) [reply swiftwinrt/code_writers/can_write.h:33]: But this PR "Fixes DEVEX-32"

**Pull request merged**